src/sagemaker_xgboost_container/data_utils.py

```diff
@@ -485,6 +483,58 @@ def get_recordio_protobuf_dmatrix(path, is_pipe=False):
         raise exc.UserError("Failed to load recordio-protobuf data with exception:\n{}".format(e))
 
 
+def _get_pipe_mode_files_path(data_path):
+    """
+     :param data_path: Either directory or file
+    """
+    # For pipe mode, we leverages mlio directly by creating a list of SageMakerPipe.
+    if isinstance(data_path, list):
+        files_path = data_path
+    else:
+        files_path = [data_path]
+        if not os.path.exists(data_path + '_0'):
+            logging.info('Pipe path {} does not exist!'.format(data_path))
+            return None
+    return files_path
+
+
+def _get_file_mode_files_path(data_path):
+    """
+     :param data_path: Either directory or file
+    """
+    # In file mode, we create a temp directory with symlink to all input files or
+    # directories to meet XGB's assumption that all files are in the same directory.
+
+    if isinstance(data_path, list):
+        logging.info('File path {} of input files'.format(data_path))
+        # Create a directory with symlinks to input files.
+        files_path = "/tmp/sagemaker_xgboost_input_data"
+        shutil.rmtree(files_path, ignore_errors=True)
+        os.mkdir(files_path)
+        for index, path in enumerate(data_path):
+            # logging for debugging remove before final merge
+
+            if not os.path.exists(path):
+                return None
+            if os.path.isfile(path):
+                base_name = os.path.join(files_path, os.path.basename(path) + '_' + str(index))
+                logging.info('creating symlink between Path {} and destination {}'.format(path, base_name))
+                os.symlink(path, base_name)
+            else:
+                for file in os.scandir(path):
+                    base_name = os.path.join(files_path, file.name + '_' + str(index))
+                    logging.info('creating symlink between Path {} and destination {}'.format(file, base_name))
+                    os.symlink(file, base_name)
```

Comment:
nit: you can factor out this duplicate code with an internal helper fun
args required are only source_path and name. Something like
```
def _make_symlink(source_path, name):
    base_name = os.path.join(files_path, name + '_' + str(index))
    logging.info('creating symlink between Path {} and destination {}'.format(source_path, base_name))
    os.symlink(source_path, base_name)
```

and then in the if bodies, you just call:
```
if os.path.isfile(path):
    _make_symlink(path, os.path.basename(path))
else:
    for file in os.scandir(path):
        _make_symlink(file, file.name)
```

Reply:
refactored